cwltool/task_queue.py

```diff
@@ -0,0 +1,119 @@
+# Copyright (C) The Arvados Authors. All rights reserved.
+#
+# SPDX-License-Identifier: Apache-2.0
+"""TaskQueue."""
+
+import queue
+import threading
+
+from typing import Callable, Optional
+
+from .loghandler import _logger
+
+
+class TaskQueue(object):
+    """
+    A TaskQueue class.
+
+    Uses a first-in, first-out queue of tasks executed on a fixed number of
+    threads.
+
+    If thread_count == 1 then as you add() tasks they will be immediately
+    executed. Otherwise new tasks enter the queue and will be executed in
+    the order received.
+
+    The thread_count is also used as the maximum size of the queue.
+
+    The threads are created during TaskQueue initialization, so delay the
+    creation of a TaskQueue until you really need it.
+
+
+    Attributes
+    ----------
+    in_flight
+        the number of tasks in the queue
+    """
+
+    def __init__(self, lock: threading.Lock, thread_count: int):
+        self.thread_count = thread_count
+        self.task_queue: queue.Queue[Optional[Callable[[], None]]] = queue.Queue(
+            maxsize=self.thread_count
+        )
+        self.task_queue_threads = []
+        self.lock = lock
+        self.in_flight = 0
+        self.error: Optional[Exception] = None
+
+        for _r in range(0, self.thread_count):
+            t = threading.Thread(target=self._task_queue_func)
+            self.task_queue_threads.append(t)
+            t.start()
+
+    def _task_queue_func(self) -> None:
```

Comment:
I made this function private, as I don't see any reason for anyone outside this class to care about it.

Reply:
Sounds good to me.